### PR TITLE
add tag specification to resources to enable security tagging

### LINKED ISSUE #10: boolean to make run_tags the same as tags
_This issue was originally opened by @nitrocode as hashicorp/packer#8695. It was migrated here as a result of the [Packer plugin split](https://github.com/hashicorp/packer/issues/8610#issuecomment-770034737). The original body of the issue is below._

<hr>

Please search the existing issues for relevant feature requests, and use the
reaction feature
(https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)
to add upvotes to pre-existing requests.

#### Feature Description

I'd like to make `run_tags` the same as `tags` using a boolean like `"tags_same" = true,` or similar which can make all tagging the same as what is defined in `tags`.

#### Use Case(s)

Tired of adding new tags to one but not the other. I'd like to keep things DRY as much as possible.

### LINKED ISSUE #35: Support tag on creation
_This issue was originally opened by @nitrocode as hashicorp/packer#10372. It was migrated here as a result of the [Packer plugin split](https://github.com/hashicorp/packer/issues/8610#issuecomment-770034737). The original body of the issue is below._

<hr>

#### Community Note

Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request.
If you are interested in working on this issue or have submitted a pull request, please leave a comment.

#### Description

https://aws.amazon.com/about-aws/whats-new/2020/12/amazon-machine-images-support-tag-on-create-tag-based-access-control/

#### Use Case(s)

Tag on creation instead of after creation

Thank you!

### LINKED ISSUE #41: AMI tag not persisting after build
_This issue was originally opened by @RixhersAjazi as hashicorp/packer#10132. It was migrated here as a result of the [Packer plugin split](https://github.com/hashicorp/packer/issues/8610#issuecomment-770034737). The original body of the issue is below._

<hr>

#### Overview of the Issue

When I set `tags` on `amazon-ebs` builder I see the tags being applied in packer log output during the running of packer. However I do not see the final tag in AWS console when I go and check on my AMI's. 

#### Reproduction Steps

- Use this amazon-ebs builder defintion:

```
    {
      "name": "t3a.micro",
      "profile": "{{user `profile`}}",
      "region": "{{user `aws_region`}}",
      "type": "amazon-ebs",
      "source_ami_filter": {
        "filters": {
          "virtualization-type": "hvm",
          "name": "ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20200609",
          "root-device-type": "ebs"
        },
        "owners": ["099720109477"],
        "most_recent": false
      },
      "ami_users": ["320176959135", "351145466100", "485438916218"],
      "instance_type": "t3a.micro",
      "ssh_username": "ubuntu",
      "ami_name": "foooo-{{timestamp}}",
      "subnet_filter": {
        "filters": {
          "tag:Group": "meta-new",
          "tag:Name": "meta-new-private-*"
        },
        "most_free": true,
        "random": false
      },
      "ssh_interface": "private_ip",
      "ami_regions": ["us-east-2", "ap-southeast-2"],
      "tags": {
        "Release": "{{user `release_tag`}}"
      },
      "run_tags":{
        "Release": "{{user `release_tag`}}"
      }
    }
```

- Run packer 
- Go to AWS CLI and see that the AMI that was generated and subsequently copied to the other regions doesn't have the tag defined:
Screen Shot 2020-10-20 at 1 09 24 PM

### Packer version

```
rixhersajazi@hostyhost packer-dev (ticketT-7772)*$ packer --version
1.6.0
```

### Packer build output

```
==> t3a.micro: Waiting for the instance to stop...
==> t3a.micro: Creating AMI fooooo-1603212597 from instance i-06aa58418cfc94449
    t3a.micro: AMI: ami-00275c4502a4e0dba
==> t3a.micro: Waiting for AMI to become ready...
==> t3a.micro: Copying/Encrypting AMI (ami-00275c4502a4e0dba) to other regions...
    t3a.micro: Copying to: us-east-2
    t3a.micro: Copying to: ap-southeast-2
    t3a.micro: Waiting for all copies to complete...
==> t3a.micro: Modifying attributes on AMI (ami-0b2881c5cba17e5b0)...
    t3a.micro: Modifying: users
==> t3a.micro: Modifying attributes on AMI (ami-00275c4502a4e0dba)...
    t3a.micro: Modifying: users
==> t3a.micro: Modifying attributes on AMI (ami-0752628f480d59636)...
    t3a.micro: Modifying: users
==> t3a.micro: Modifying attributes on snapshot (snap-04a99f316e19fced0)...
==> t3a.micro: Modifying attributes on snapshot (snap-032d3d88e4844f1dd)...
==> t3a.micro: Modifying attributes on snapshot (snap-0bd7cae46477fc6ca)...
==> t3a.micro: Adding tags to AMI (ami-00275c4502a4e0dba)...
==> t3a.micro: Tagging snapshot: snap-04a99f316e19fced0
==> t3a.micro: Creating AMI tags
    t3a.micro: Adding tag: "Release": "FEATURE_BRANCH_QA_ONLY"
==> t3a.micro: Creating snapshot tags
==> t3a.micro: Adding tags to AMI (ami-0752628f480d59636)...
==> t3a.micro: Tagging snapshot: snap-032d3d88e4844f1dd
==> t3a.micro: Creating AMI tags
    t3a.micro: Adding tag: "Release": "FEATURE_BRANCH_QA_ONLY"
==> t3a.micro: Creating snapshot tags
==> t3a.micro: Adding tags to AMI (ami-0b2881c5cba17e5b0)...
==> t3a.micro: Tagging snapshot: snap-0bd7cae46477fc6ca
==> t3a.micro: Creating AMI tags
    t3a.micro: Adding tag: "Release": "FEATURE_BRANCH_QA_ONLY"
==> t3a.micro: Creating snapshot tags
==> t3a.micro: Terminating the source AWS instance...
==> t3a.micro: Cleaning up any extra volumes...
==> t3a.micro: No volumes to clean up, skipping
==> t3a.micro: Deleting temporary security group...
==> t3a.micro: Deleting temporary keypair...
Build 't3a.micro' finished.

==> Builds finished. The artifacts of successful builds are:
--> t3a.micro: AMIs were created:
ap-southeast-2: ami-0752628f480d59636
us-east-2: ami-0b2881c5cba17e5b0
us-west-2: ami-00275c4502a4e0dba

```

Notice there are lines saying: 

```
==> t3a.micro: Creating AMI tags
    t3a.micro: Adding tag: "Release": "FEATURE_BRANCH_QA_ONLY"
```

### Operating system and Environment details

MacOS Catalina - 10.15.7

### LINKED ISSUE #44: amazon-ebs: fine-grained IAM policy
_This issue was originally opened by @JIoJIaJIu as hashicorp/packer#10092. It was migrated here as a result of the [Packer plugin split](https://github.com/hashicorp/packer/issues/8610#issuecomment-770034737). The original body of the issue is below._

<hr>

#### Description
I would like to be able to create a custom IAM policy that grants access to the Packer, and only to the resources created by Packer. AWS uses conditions base on `aws:RequestTag` and `aws:ResourceTag` for that purpose, but unfortunately for most of the controlled resources Packer doesn't setup the tags. It leads that CI/CD user has wide permissions - ability to remove/change resources it shouldn't. At this moment I found the only way to limit access by using separated aws region for the Packer

Similar issues: hashicorp/packer#9894 

#### Use Case(s) / Potential configuration
So would be great:
1. <details><summary>:pencil2: Apply tag `Managed by: Hashicor Packer` to the all existed resources that support it</summary>

    It will allow writing policy as
    ```hcl
    ...
    statement {
      actions = [
        "ec2:CreateSecurityGroup",
      ]
      resources = ["*"]
      condition {
        test  = "StringEquals"
        variable = "aws:RequestTag/Managed by"
        values = [
          "Hashicorp Packer"
        ]
      }
    }

    statement {
      actions = [
        "ec2:AuthorizeSecurityGroupIngress",
        "ec2:DeleteSecurityGroup",
      ]
      resources = ["*"]
      condition {
        test  = "StringEquals"
        variable = "aws:ResourceTag/Managed by"
        values = [
          "Hashicorp Packer"
        ]
      }
    }
    ...
    ```
    
    For actions that supports resources it's possible to point like
    ```hcl
    statement {
        actions = [
          "ec2:CreateKeyPair",
          "ec2:DeleteKeyPair",
        ]
 
        resources = ["arn:aws:ec2:*:${var.aws_account}:key-pair/packer_*"]
     }
    ```
    but 
    1. it's not supported by most of the actions
    2. using tags will make policy simpler and more uniform 
    </details>
2. <details><summary>:bell: Add base tag during creation with TagSpecification for the resources</summary>

    ```go
    runOpts := &ec2.RunInstancesInput{
      ImageId:             &s.SourceAMI,
      ...
      TagSpecifications: []*ec2.TagSpecification{
        {
          ResourceType: aws.String("instance"),
          Tags: []*ec2.Tag{
            {
              Key: aws.String("Managed by"),
              Value: aws.String("Hashicorp Packer"),
            },
          },
        },
        ...
      },
    }
    ```
    to be able to create policy
    ```json
    {
      "Effect": "Allow",
      "Action": [
        "ec2:RunInstances"
      ],
      "Resource": [
        "arn:aws:ec2:*:${var.aws_account}:instance/*"
      ],
      "Condition": {
        "StringEquals": {
          "aws:RequestTag/Managed by": "Hashicorp Packer"
        }
      }
     ```
     rather than apply it after on the created instance, cause it requires the policy that allows to add any tag
    </details>
3. <details><summary>:closed_book: Update minimal IAM policy in the doc</summary>

    It would be great if IAM policy [is provided in docs](https://www.packer.io/docs/builders/amazon#iam-task-or-instance-role) contains these conditions.
    </details>

#### Potential References
* https://docs.aws.amazon.com/IAM/latest/UserGuide/list_amazonec2.html
* https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/control-access-with-tags.html
* https://aws.amazon.com/blogs/security/simplify-granting-access-to-your-aws-resources-by-using-tags-on-aws-iam-users-and-roles/

### LINKED ISSUE #45: amazon-ebs: Temporary snapshot not tagged causing permission denied
_This issue was originally opened by @rednuht as hashicorp/packer#9894. It was migrated here as a result of the [Packer plugin split](https://github.com/hashicorp/packer/issues/8610#issuecomment-770034737). The original body of the issue is below._

<hr>

#### Overview of the Issue

In our company we use tags to dictate whether a role/user has access to perform actions on a resource.
When packer tries to clean up at the end it is unable to delete the temporary snapshot since it is untagged.

#### Reproduction Steps

Include a policy statement that has a condition on tags:

        {
            "Effect": "Allow",
            "Action": "ec2:*",
            "Resource": "*",
            "Condition": {
                "StringEquals": {
                    "ec2:ResourceTag/Team": "MY_SPECIAL_TAG"
                }
            }
        }

### Packer version

From `1.3.2` and onwards.
(haven't tested with earlier versions)

### Simplified Packer Buildfile

https://gist.github.com/rednuht/f2c62fbc47186e41d0f8e9707954bcc7

### Operating system and Environment details

Not relevant.

### Log Fragments and crash.log files

    2020-09-04T09:44:21+02:00: Build 'amazon-ebs' errored after 52 minutes 56 seconds: Error deleting existing snapshot: retry count exhausted. Last err: UnauthorizedOperation: You are not authorized to perform this operation. Encoded authorization failure message: ENCODED MSG status code: 403, request id: 6b6df0be-3b0d-4eb7-8185-41823465f67d

Important parts from decoded message:

```
{
  "allowed": false,
  "explicitDeny": false,
  "matchedStatements": {
    "items": []
  },
  "failures": {
    "items": []
  },
  "context": {
    "action": "ec2:DeleteSnapshot",
    "resource": "arn:aws:ec2:eu-west-1::snapshot/snap-FOOBAR123",
  }
}
```

### LINKED ISSUE #62: Add tags to temporary security groups (second attempt)
_This issue was originally opened by @vsevolod-fedorov as hashicorp/packer#10885. It was migrated here as a result of the [Packer plugin split](https://github.com/hashicorp/packer/issues/8610#issuecomment-770034737). The original body of the issue is below._

<hr>

A am composing IAM role with aim to limit access for packer to resources it does not create. In particular, I want to prevent it from deleting security groups it did not create.
Unfortunately, I have no means to limit this - packer does not have feature to add tags to security groups it creates, and security group name (packer_*) is not mapped to any of resource filters available for IAM to limit on.

So I would suggest to add element like 'security_group_tags' to amazon source.

We have similar, but slightly different closed issue:  hashicorp/packer#2859

[ec2 resources](https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazonec2.html#amazonec2-resources-for-iam-policies)
[available security group resource filters](https://github.com/hashicorp/packer/files/6276436/sg-resource-fields.txt)

### PR DESCRIPTION
**Pull request by wolffberg**
The EC2 service allows for most resources to be tagged on creation. This, together with the [CreateAction IAM Condition](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/supported-iam-actions-tagging.html), allows us to restrict Packer to manage only the resources it creates.

This PR adds [TagSpecification](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_TagSpecification.html)s to the following resources using the `run_tags` variable.

- Key-pairs
- Security Groups
- EBS based Instances
- EBS based AMIs
- EBS based Snapshots

The `run_tags` variable was chosen to avoid clutter in the `source` block but as the tags are now attached to both the AMI and snapshot after Packer has finished running one could argue that they should be moved to their own variable.

The tags added on creation of key-pairs, security groups, AMIs and snapshots are currently not outputted to the console which might be desirable.

EC2 Instance tags will be outputted to the console and might show up twice as they are added on creation and added after creation. This was not changed to preserve backwards compatibility. A future PR could migrate all tags to  [TagSpecification](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_TagSpecification.html)s.

The plugin has been tested both with and without `run_tags` but I have not been able to test the China region restrictions as I don't have access to the regions. [CloudGov](https://aws.amazon.com/about-aws/whats-new/2018/07/tag-on-create-for-amazon-ec2-and-amazon-ebs-is-now-available-in-the-aws-govcloud-region/) regions are [almost](https://github.com/hashicorp/packer-plugin-amazon/pull/96#issuecomment-870358088) fully supported.

Closes #10 (possible side effect)
Closes #35
Closes #41 (possibly)
Closes #44
Closes #45
Closes #62

A restrictive IAM policy could look like this

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Describe",
            "Effect": "Allow",
            "Action": "ec2:Describe*",
            "Resource": "*"
        },
        {
            "Sid": "Create",
            "Effect": "Allow",
            "Action": [
                "ec2:CreateKeyPair",
                "ec2:CreateSecurityGroup",
                "ec2:RunInstances",
                "ec2:CreateSnapshot",
                "ec2:CreateImage"
            ],
            "Resource": "*"
        },
        {
            "Sid": "Tag",
            "Effect": "Allow",
            "Action": "ec2:CreateTags",
            "Resource": "*",
            "Condition": {
                "StringEquals": {
                    "ec2:CreateAction": [
                        "CreateKeyPair",
                        "CreateSecurityGroup",
                        "RunInstances",
                        "CreateSnapshot",
                        "CreateImage"
                    ]
                }
            }
        },
        {
            "Sid": "Manage",
            "Effect": "Allow",
            "Action": "ec2:*",
            "Resource": "*",
            "Condition": {
                "StringEquals": {
                    "aws:ResourceTag/PackerSec": [
                        "Packer"
                    ]
                }
            }
        }
    ]
}
```

### CODE REVIEW

**Review by SwampDragons — CHANGES_REQUESTED** [review]
Hi, this looks good. I think it should be okay to just use run_tags for these resources, but we need to update docs. To do that, you'll want to update the docstring here https://github.com/hashicorp/packer-plugin-amazon/blob/main/builder/common/run_config.go#L212 and then regenerate docs by calling `make generate` in the plugin root. 

I don't have the ability to test the aws-china restrictions either -- do you have a link to any kind of documentation that explains why you special-cased China like that? Does that region not support this kind of tagging?

**Comment by wolffberg** [comment]
Updated docs 📃

I found [this](https://aws.amazon.com/about-aws/whats-new/2018/07/tag-on-create-for-amazon-ec2-and-amazon-ebs-is-now-available-in-the-aws-govcloud-region/) article about the feature becoming available on GovCloud but didn't find any articles explicitly saying that the feature were available in China regions so I didn't want to take the chance 😕

**Comment by wolffberg** [comment]
Could use a second opinion.
Reading [this](https://aws.amazon.com/about-aws/whats-new/2020/12/amazon-machine-images-support-tag-on-create-tag-based-access-control/) article it could look like you cannot tag AMI's on creation in both GovCloud and China regions. If that's the case we need to add `IsGovCloud()` to `stepCreateAMI`.
I'm not sure how many restrictions to add without being able to test myself.

**Comment by SwampDragons** [comment]
Ah, thanks for the link. I totally missed that line on my first read through.

**Comment by SwampDragons** [comment]
And yes, you probably want to add the govcloud exception too based on that article. I'll try to finish reviewing the PR this week. I've been swamped, and since this does change default behavior I want to spend some more time thinking over it before I am comfortable mergeing.

**Comment by wolffberg** [comment]
Updated with the new restriction.

And thanks. My use case is pretty simple so testing more complex cases would probably be best 🎉

**Comment by aston-r** [comment]
@SwampDragons 

> I'll try to finish reviewing the PR this week. 

Sorry for bothering but I am also waiting for this new functionality. Already a month has passed. Is there any chance that you review and merge the changes in the nearest time?

File: builder/common/step_key_pair.go
Comment (azr) [inline builder/common/step_key_pair.go:69]:
```suggestion
	if !s.IsRestricted {
```

Comment (azr) [reply builder/common/step_key_pair.go:69]: (trying to commit to see if the tests are triggered)

**Review by azr — APPROVED** [review]
This LGTM, would like to leave the final call to @SwampDragons as I don't have much exp in this area 🙂

**Comment by limasilet** [comment]
Hi! We are also looking forward to this PR content! Is there any chance this could be merged soon?

**Comment by rquadling** [comment]
@limasilet I think https://www.theregister.com/2021/09/07/hashicorp_pause/ may explain things a little.

**Comment by limasilet** [comment]
> @limasilet I think https://www.theregister.com/2021/09/07/hashicorp_pause/ may explain things a little.

@rquadling I was really afraid of that. But kept hoping it would affect only Terraform :(

**Review by nywilken — CHANGES_REQUESTED** [review]
Hi there folks, thanks for keeping a pulse on this PR. We are actively reviewing PRs but have been focusing on the HCL UX, plugin split clean up, along with adding functionality for HCP packer. Thanks for your understanding if feedback has been slower than anticipated.

I reviewed this PR and left a few suggestions that we should aim to address before merging. To @SwampDragons point we would like to give this some testing since it does change default behavior.

**Comment by nywilken** [comment]
For folks following the PR, if you are able to help with testing you could run one of the [test binaries](https://app.circleci.com/pipelines/github/hashicorp/packer-plugin-amazon/48/workflows/1fdc7a00-42be-40bb-bb26-afefcacd7eb7/jobs/607/artifacts) against your own testing environment/configuration to validated the expected behavior. Thanks in advance.

File: builder/common/step_key_pair.go
Comment (wolffberg) [reply builder/common/step_key_pair.go:22]: While I agree this could be changed, the name is used throughout Packer.
I suggest we keep it as is in this PR and make a new one later, changing all the occurences to something like `IsRegionRestricted` instead.
What do you think?

File: builder/common/run_config.go
Comment (wolffberg) [inline builder/common/run_config.go:213]:
```suggestion
	// that is *launched* to create the EBS volumes. The resulting AMI will also inherit these tags.
```

**Comment by limasilet** [comment]
> For folks following the PR, if you are able to help with testing you could run one of the [test binaries](https://app.circleci.com/pipelines/github/hashicorp/packer-plugin-amazon/48/workflows/1fdc7a00-42be-40bb-bb26-afefcacd7eb7/jobs/607/artifacts) against your own testing environment/configuration to validate the expected behavior. Thanks in advance.

I tried testing it, but it's still not working for me. Would you have instructions on how to place the plugin binaries in my local env for testing? I tried all the documented ways I could find, but the result is the same. :( 
```==> aws: InvalidTagKey.Malformed: Tag specification must have at least one tag
==> aws:        status code: 400, request id: c950e354-d9e1-45a2-9696-c0284e4e02cc```

File: builder/common/step_key_pair.go
Comment (nywilken) [reply builder/common/step_key_pair.go:22]: Sounds like a good way forward to me. Thanks for thinking a little deeper on this suggestion.

**Comment by nywilken** [comment]
I found this document https://docs.amazonaws.cn/en_us/AWSEC2/latest/UserGuide/Using_Tags.html which seems to imply that there is support for tagging on creation in the China Region, but as you said without a test we can't confirm. I reached out internally to see if there is prior art to this. But I think moving forward as it is now is good, and we can come back an update to allow China region if requested by a user who can test.

**Comment by nywilken** [comment]
> > For folks following the PR, if you are able to help with testing you could run one of the [test binaries](https://app.circleci.com/pipelines/github/hashicorp/packer-plugin-amazon/48/workflows/1fdc7a00-42be-40bb-bb26-afefcacd7eb7/jobs/607/artifacts) against your own testing environment/configuration to validate the expected behavior. Thanks in advance.
> 
> @nywilken Hi! I tried testing it, but it's still not working for me. Would you have instructions on how to place the plugin binaries in my local env for testing? I tried all the documented ways I could find, but the result is the same. :(
> 
> ```
> ==> aws: InvalidTagKey.Malformed: Tag specification must have at least one tag
> ==> aws:        status code: 400, request id: c950e354-d9e1-45a2-9696-c0284e4e02cc 
> ```

@limasilet thanks for trying to test. Since this is not an official release of the plugin it requires a little extra work on the template file to use the built binary. Please let me know if the following steps work for you. 

1. Download the binary from [here](https://app.circleci.com/pipelines/github/hashicorp/packer-plugin-amazon/48/workflows/1fdc7a00-42be-40bb-bb26-afefcacd7eb7/jobs/607/artifacts)
2. Unzip the downloaded binary into ~/.packer.d/plugins/ 
3. Remove previously installed version of the Amazon binary `rm -r ~/.packer.d/plugins/github.com/hashicorp/amazon`
4. Remove amazon from the required_plugins block in your template if you are using `packer init` to install the plugin. 
```
 packer {
   required_plugins {
-    amazon = {
-      version = ">= 0.0.1"
-      source = "github.com/hashicorp/amazon"
-    }
   }
 }
```
5. Run packer build on the updated template which should use the test binary for Amazon. 

If you are running into the same issue after taking the steps above please share a redacted version of your template file so that we can see if it is an issue with the template or the code changes. So far the change looks good in my testing.

**Comment by limasilet** [comment]
> > > For folks following the PR, if you are able to help with testing you could run one of the [test binaries](https://app.circleci.com/pipelines/github/hashicorp/packer-plugin-amazon/48/workflows/1fdc7a00-42be-40bb-bb26-afefcacd7eb7/jobs/607/artifacts) against your own testing environment/configuration to validate the expected behavior. Thanks in advance.
> > 
> > 
> > @nywilken Hi! I tried testing it, but it's still not working for me. Would you have instructions on how to place the plugin binaries in my local env for testing? I tried all the documented ways I could find, but the result is the same. :(
> > ```
> > ==> aws: InvalidTagKey.Malformed: Tag specification must have at least one tag
> > ==> aws:        status code: 400, request id: c950e354-d9e1-45a2-9696-c0284e4e02cc 
> > ```
> 
> @limasilet thanks for trying to test. Since this is not an official release of the plugin it requires a little extra work on the template file to use the built binary. Please let me know if the following steps work for you.
> 
> 1. Download the binary from [here](https://app.circleci.com/pipelines/github/hashicorp/packer-plugin-amazon/48/workflows/1fdc7a00-42be-40bb-bb26-afefcacd7eb7/jobs/607/artifacts)
> 2. Unzip the downloaded binary into ~/.packer.d/plugins/
> 3. Remove previously installed version of the Amazon binary `rm -r ~/.packer.d/plugins/github.com/hashicorp/amazon`
> 4. Remove amazon from the required_plugins block in your template if you are using `packer init` to install the plugin.
> 
> ```
>  packer {
>    required_plugins {
> -    amazon = {
> -      version = ">= 0.0.1"
> -      source = "github.com/hashicorp/amazon"
> -    }
>    }
>  }
> ```
> 
> 1. Run packer build on the updated template which should use the test binary for Amazon.
> 
> If you are running into the same issue after taking the steps above please share a redacted version of your template file so that we can see if it is an issue with the template or the code changes. So far the change looks good in my testing.

@nywilken  Thank you for the instructions! What do I do if I can't use packer init? Not using HCL and can't convert at this time :(

**Review by nywilken — APPROVED** [review]
LGTM

**Comment by wolffberg** [comment]
Sorry to hear that. I was a bit unsure if it would solve the issue (hence "possibly").
@nywilken can you reopen #41?

**Pull request merged**
